pkg/processor/worker.go

```diff
@@ -62,25 +63,46 @@ func (wrk *worker) rebuild(bundle *smith.Bundle) error {
 	}
 	log.Printf("[WORKER] Rebuilding the bundle %s/%s", wrk.namespace, wrk.bundleName)
 
+	// Build resource map by name
+	resourceMap := make(map[string]smith.Resource)
 	for _, res := range bundle.Spec.Resources {
-		if wrk.checkNeedsRebuild() {
-			return nil
-		}
-		resClone, err := wrk.tp.scheme.DeepCopy(&res)
-		if err != nil {
-			return err
-		}
-		isReady, err := wrk.checkResource(bundle, resClone.(*smith.Resource))
-		if err != nil {
-			return err
-		}
-		if !isReady {
-			if err := wrk.setBundleState(bundle, smith.IN_PROGRESS); err != nil {
-				log.Printf("[WORKER] %v", err)
+		resourceMap[res.Name] = res
+	}
+
+	log.Printf("Sorting resources")
+	// Split resources into dependency graphs and topologically sort them
+	graphs, err := graph.TopologicalSort(bundle)
+	if err != nil {
+		return err
+	}
+
+	// Visit each dependency subgraph
+	for _, names := range graphs {
+		// Visit resources of each dependency subgraph in topologically sorted order
+		for _, name := range names {
+			log.Printf("Checking resource %s", name)
+			res := resourceMap[name]
+
+			if wrk.checkNeedsRebuild() {
+				return nil
+			}
+			resClone, err := wrk.tp.scheme.DeepCopy(&res)
+			if err != nil {
+				return err
+			}
+			isReady, err := wrk.checkResource(bundle, resClone.(*smith.Resource))
+			if err != nil {
+				return err
+			}
+			if !isReady {
+				if err := wrk.setBundleState(bundle, smith.IN_PROGRESS); err != nil {
+					log.Printf("[WORKER] %v", err)
+				}
+				break
```

Comment:
If we found a not-ready-yet resource, break from the inner for loop and move to the next graph